README.md
# fule_swiper
![Pub Version](https://img.shields.io/pub/v/fule_swiper?style=flat-square)
![GitHub](https://img.shields.io/github/license/Yang-Xuhao/fule_swiper?style=flat-square)

一个简单的Swiper组件，它允许你在其中放置任意的组件。

<img src="https://i.loli.net/2020/04/16/9Pi7Tb81AqjQFho.gif"  width="40%"/>

## 开始
### 安装
在pubspec.yaml中添加引用
``` yaml
fule_swiper: ^0.1.2
```
下载插件
```shell
$ flutter pub get
```
### 引入
``` dart
import 'package:fule_swiper/fule_swiper.dart';
```
### 上面演示的源码
``` dart
import 'package:flutter/material.dart';
import 'package:fule_swiper/fule_swiper.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  MyApp({Key key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List images = [
    {
      "src": "https://i.loli.net/2020/04/16/gpIVeN8a2m93nXt.jpg",
      "desc": "小船在水上"
    },
    {
      "src": "https://i.loli.net/2020/04/16/zRf7qoAlJV6kIcN.jpg",
      "desc": "冰雪覆盖"
    },
    {
      "src": "https://i.loli.net/2020/04/16/eQx6rlmXNiwUzPO.jpg",
      "desc": "绿色中的棕房子"
    },
  ];

  List<Widget> getImages() {
    List<Widget> _list = List();
    images.forEach((_item) {
      _list.add(Stack(
        children: <Widget>[
          Image.network(
            _item['src'],
            fit: BoxFit.fitHeight,
            width: 750,
            height: 1334,
          ),
          Align(
            alignment: Alignment(-0.9, 0.3),
            child: Text(
              _item['desc'],
              style: TextStyle(
                color: Colors.white,
                fontSize: 48,
              ),
            ),
          )
        ],
      ));
    });
    return _list;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Swiper(
            width: 750,
            height: 1334,
            autoplay: true,
            duration: 500,
            interval: 2000,
            indicatorActiveColor: Colors.white,
            indicatorColor: Colors.white38,
            indicatorRadius: 4,
            indicatorSpacing: 16,
            indicatorAlignment: Alignment(0, 0.95),
            children: getImages(),
          ),
        ),
      ),
    );
  }
}
```
### 参数
| 属性 | 类型 | 默认值 | 必填 | 说明 | 组件版本 |
| --- | --- | --- | :---: | --- | :---: |
| width| double | 100.0 | 否 | 组件的宽度 | 0.1.0 |
| height | double | 100.0 | 否 | 组件的高度 | 0.1.0 |
| children | List&lt;Widget&gt; | - | 是 | 子组件 | 0.1.0 |
| current | int | 0 | 否 | 默认选中的下标 | 0.1.0 |
| autoplay | bool | false | 否 | 是否自动切换 | 0.1.0 |
| interval | int | 2000 | 否 | 自动切换时间间隔 | 0.1.0 |
| duration | int | 1000 | 否 | 滑动动画时长 | 0.1.0 |
| curve | Curve | Curves.ease | 否 | 滑动动画 | 0.1.0 |
| indicatorDot | bool | true | 否 | 是否显示指示点 | 0.1.2 |
| indicatorColor | Color | Colors.white | 否 | 指示点默认颜色 | 0.1.0 |
| indicatorActiveColor | Color | Colors.white | 否 | 指示点选中颜色 | 0.1.0 |
| indicatorRadius | double | 4.0 | 否 | 指示点半径 | 0.1.0 |
| indicatorSpacing | double | 8.0 | 否 | 指示点间距 | 0.1.0 |
| indicatorAlignment | Alignment | Alignment.bottomCenter | 否 | 指示点的位置 | 0.1.2 |
| onPageChanged | bool | - | 否 | 切换回调函数 | 0.1.0 |

## TODO
- [ ] 实现纵向滑动
- [ ] 实现循环播放

CHANGELOG.md
## [0.1.2] 2020-04-16

* 修复indicatorDot无效的问题
* 添加indicatorAlignment控制指示点位置
* 更新了例子

## [0.1.1] 2020-04-13

* 把上次的CHANGELOG补上
* 把更新了README

## [0.1.0] 2020-04-13

* 完善pubspec.yaml中的description
* 把README中的样例放到了example目录下
* 改了个版本号

## [0.0.1] 2020-04-11

* Swiper第一个版本

test/fule_swiper_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:fule_swiper/fule_swiper.dart';

void main() {

}


lib/src/swiper.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'dot_sequence.dart';

class Swiper extends StatefulWidget {
  final double width;
  final double height;
  final List<Widget> children;
  final int current;
  final bool autoplay;
  final int duration;
  final int interval;
  final Curve curve;
  final bool indicatorDot;
  final Color indicatorColor;
  final Color indicatorActiveColor;
  final double indicatorRadius;
  final double indicatorSpacing;
  final Alignment indicatorAlignment;
  final ValueChanged<int> onPageChanged;

  Swiper({
    Key key,
    this.width = 100.0,
    this.height = 100.0,
    @required this.children,
    this.current = 0,
    this.autoplay = false,
    this.duration = 500,
    this.interval = 4000,
    this.curve = Curves.ease,
    this.indicatorDot = true,
    this.indicatorColor = Colors.white,
    this.indicatorActiveColor = Colors.blue,
    this.indicatorRadius = 4.0,
    this.indicatorSpacing = 8.0,
    this.indicatorAlignment = Alignment.bottomCenter,
    this.onPageChanged,
  })  : assert(children != null),
        assert(duration <= interval),
        assert(current >= 0 && current < children.length),
        super(key: key);

  @override
  _SwiperState createState() => _SwiperState(this.current, this.children);
}

class _SwiperState extends State<Swiper> {
  int itemCount;
  int _index;
  SliverChildDelegate _sliverChildDelegate;
  PageController _pageController;
  List<Timer> _timerList = List();

  _SwiperState(current, children) {
    //创建delegate
    this._sliverChildDelegate = SliverChildListDelegate(children);
    //设置显示的页面序号
    this._index = current;
    //创建controller
    this._pageController = PageController(initialPage: this._index);
  }

  @override
  Widget build(BuildContext context) {
    //创建定时器
    if (widget.autoplay) {
      if (this._timerList.length != 0) {
        // print('TimerList数量${this._timerList.length}');
        this._timerList.forEach((timer) {
          if (timer.isActive) {
            timer.cancel();
          }
        });
        this._timerList.clear();
      }

      this
          ._timerList
          .add(Timer.periodic(Duration(milliseconds: widget.interval), (timer) {
            if (this._index == widget.children.length - 1) {
              this._pageController.animateToPage(
                    0,
                    duration: Duration(milliseconds: widget.duration),
                    curve: widget.curve,
                  );
            } else {
              this._pageController.nextPage(
                    duration: Duration(milliseconds: widget.duration),
                    curve: widget.curve,
                  );
            }
            this._timerList.last.cancel();
          }));
    }

    //是否显示指示器
    Widget getIndicatorDot() {
      if (widget.indicatorDot) {
        return Align(
          alignment: widget.indicatorAlignment,
          child: DotSequence(
            radius: widget.indicatorRadius,
            spacing: widget.indicatorSpacing,
            color: widget.indicatorColor,
            activeColor: widget.indicatorActiveColor,
            current: this._index,
            itemCount: widget.children.length,
          ),
        );
      } else {
        return SizedBox();
      }
    }

    return Container(
      color: Colors.transparent,
      width: widget.width,
      height: widget.height,
      child: Stack(
        children: <Widget>[
          PageView.custom(
              childrenDelegate: this._sliverChildDelegate,
              controller: this._pageController,
              onPageChanged: (index) {
                setState(() {
                  this._index = index;
                  widget.onPageChanged?.call(index);
                });
              }),
          getIndicatorDot()
        ],
      ),
    );
  }
}


lib/src/dot_sequence.dart
import 'package:flutter/material.dart';

class DotSequence extends StatelessWidget {
  final double radius;
  final double spacing;
  final Color color;
  final Color activeColor;
  final int current;
  final int itemCount;
  DotSequence({
    Key key,
    this.radius,
    this.spacing,
    this.color,
    this.activeColor,
    this.current = 0,
    this.itemCount = 0,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    //定义normal圆点
    Widget _normalDot = Container(
      height: radius * 2,
      width: radius * 2,
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        color: color,
      ),
    );
    //定义active圆点
    Widget _activeDot = Container(
      height: radius * 2,
      width: radius * 2,
      decoration: BoxDecoration(
        shape: BoxShape.circle,
        color: activeColor,
      ),
    );
    //定义间距
    Widget _spacing = SizedBox(
      height: radius * 2,
      width: spacing,
    );
    //生成圆点列表
    List<Widget> _dotList = List();
    for (var i = 0; i < itemCount; i++) {
      i == current ? _dotList.add(_activeDot) : _dotList.add(_normalDot);
      if (i != itemCount - 1) {
        _dotList.add(_spacing);
      }
    }

    return Row(
      mainAxisSize: MainAxisSize.min,
      children: _dotList,
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:fule_swiper/fule_swiper.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  MyApp({Key key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List images = [
    {
      "src": "https://i.loli.net/2020/04/16/gpIVeN8a2m93nXt.jpg",
      "desc": "小船在水上"
    },
    {
      "src": "https://i.loli.net/2020/04/16/zRf7qoAlJV6kIcN.jpg",
      "desc": "冰雪覆盖"
    },
    {
      "src": "https://i.loli.net/2020/04/16/eQx6rlmXNiwUzPO.jpg",
      "desc": "绿色中的棕房子"
    },
  ];

  List<Widget> getImages() {
    List<Widget> _list = List();
    images.forEach((_item) {
      _list.add(Stack(
        children: <Widget>[
          Image.network(
            _item['src'],
            fit: BoxFit.fitHeight,
            width: 750,
            height: 1334,
          ),
          Align(
            alignment: Alignment(-0.9, 0.3),
            child: Text(
              _item['desc'],
              style: TextStyle(
                color: Colors.white,
                fontSize: 48,
              ),
            ),
          )
        ],
      ));
    });
    return _list;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Swiper(
            width: 750,
            height: 1334,
            autoplay: true,
            duration: 500,
            interval: 2000,
            indicatorActiveColor: Colors.white,
            indicatorColor: Colors.white38,
            indicatorRadius: 4,
            indicatorSpacing: 16,
            indicatorAlignment: Alignment(0, 0.95),
            children: getImages(),
          ),
        ),
      ),
    );
  }
}
